Li thresholding port from imagej

jmetz opened the pull request:
Ported imagej version of Li thresholding (see refs in docstring).

jmetz commented:
**Note:** Travis build is failing in unrelated areas, e.g. bottom of https://travis-ci.org/scikit-image/scikit-image/jobs/51085857 shows

    loading pickled environment... not yet created
    loading intersphinx inventory from http://scikit-learn.org/stable/objects.inv...
    Warning, treated as error:
    WARNING: intersphinx inventory 'http://scikit-learn.org/stable/objects.inv' not fetchable due to <class 'socket.error'>: [Errno 104] Connection reset by peer
    make[1]: *** [html] Error 1
    make[1]: Leaving directory `/home/travis/build/scikit-image/scikit-image/doc'
    make: *** [html] Error 2
    The command "tools/travis_script.sh" exited with 2.
    Done. Your build exited with 1.

blink1073 commented on skimage/filters/thresholding.py:
```diff
@@ -302,3 +303,68 @@ def threshold_isodata(image, nbins=256, return_all=False):
+    Returns
+    -------
+    threshold : float
+        Upper threshold value. All pixels intensities that less or equal of
```
that are less

jmetz replied:
Copied the description from the IJ code

blink1073 replied:
I'd still argue for changing it.  :smile:

stefanv replied:
Err, please fix this.

blink1073 commented on skimage/filters/thresholding.py:
```diff
@@ -302,3 +303,68 @@ def threshold_isodata(image, nbins=256, return_all=False):
+    >>> binary = image <= thresh
+    """
+
+    tolerance=0.5
```
pep8: add spaces between the `=`

blink1073 commented on skimage/filters/thresholding.py:
```diff
@@ -302,3 +303,68 @@ def threshold_isodata(image, nbins=256, return_all=False):
+    
+    # Initial estimate 
+    new_thresh = mean
+    old_thresh = new_thresh + 2*tolerance
```
add spaces between `*`

blink1073 commented on skimage/filters/thresholding.py:
```diff
@@ -302,3 +303,68 @@ def threshold_isodata(image, nbins=256, return_all=False):
+
+        temp = (mean_back - mean_obj) / (np.log(mean_back) - np.log(mean_obj))
+
+        if temp < 0: # (temp < -2.220446049250313E-16)
```
Why is the commented value there?  Was that in the ImageJ version and no longer needed?

jmetz replied:
Thx for the feedback @blink1073  - will fix the PEP8 issues asap. 

See above about the IJ description (re: "that are less" ).

The value is from the IJ version - left it in to get confirmation that we def. don't need to handle floating point issues explicitly..?

blink1073 replied:
We typically don't test against an `epsilon`, so I'd say remove the comment.

Also, can you run a "remove trailing whitespace" in your editor?

blink1073 commented:
@jmetz, thanks for the contribution!  That error looks like a temporary connectivity issue.  
Can you please add a test for this in [test_thresholding](https://github.com/scikit-image/scikit-image/blob/master/skimage/filters/tests/test_thresholding.py) based on one of the other tests?

blink1073 commented on skimage/filters/thresholding.py:
```diff
@@ -302,3 +303,68 @@ def threshold_isodata(image, nbins=256, return_all=False):
+
+    # Stop the iterations when the difference between the
+    # new and old threshold values is less than the tolerance 
+    while abs( new_thresh - old_thresh ) > tolerance: 
```
Pep8: please remove all spaces between `(` and `[` (also below).

jmetz replied:
@blink1073 : How are the tests created in `test_thresholding`? 
E.g. the `test_otsu_coins_image`
    
    def test_otsu_lena_image():
        img = skimage.img_as_ubyte(data.lena())
        assert 140 < threshold_otsu(img) < 142

-- was the range (140--142) determined using an external implementation?

blink1073 replied:
@jmetz, typically we have run the algorithms against one of our test images, verified that the image is visually correct and then used the resulting array as a regression test.

jmetz replied:
@blink1073 - thanks for the clarification - I'll run the algorithm on some of the test images and add in the tests over the next few days.

stefanv commented:
If this is a line-by-line translation, we should include their copyright in the local directory's README.txt.

jmetz commented:
@stefanv : It's quite a heavy adaptation (runs on image instead of histogram, replaced large portions original with one-line `numpy` equivalents) - guess it's a bit grey as I used the IJ code as a "pseudo-code" template.

ahojnnes commented on skimage/filters/thresholding.py:
```diff
@@ -302,3 +303,68 @@ def threshold_isodata(image, nbins=256, return_all=False):
+        # Background 
+        mean_back = image[ image <= threshold ].mean()
+        # Object 
+        mean_obj = image[ image > threshold ].mean()
```
Remove spaces around []

jmetz commented:
@blink1073 : Added tests (using visual checks for consistency) - also added licence file as suggested by @stefanv. Updated algorithm to subtract  (then finally re-add) `min` to handle negative images.

blink1073 commented on skimage/filters/tests/test_thresholding.py:
```diff
@@ -152,6 +164,29 @@ def test_otsu_astro_image():
+    assert 0.37 < threshold_li(coins) < 0.38
+
+
+def test_li_lena_image():
```
Please remove the lena example, we're slowing removing lena from scikit-image (due to the provocative nature of the image).

blink1073 commented on skimage/filters/thresholding.py:
```diff
@@ -302,3 +303,74 @@ def threshold_isodata(image, nbins=256, return_all=False):
+    .. [1] Li C.H. and Lee C.K. (1993) "Minimum Cross Entropy Thresholding"
+           Pattern Recognition, 26(4): 617-625
+    .. [2] Li C.H. and Tam P.K.S. (1998) "An Iterative Algorithm for Minimum
+           Cross Entropy Thresholding"Pattern Recognition Letters, 18(8): 771-776
```
Missing space after quotation mark.

blink1073 commented on skimage/filters/thresholding.py:
```diff
@@ -302,3 +303,74 @@ def threshold_isodata(image, nbins=256, return_all=False):
+    >>> thresh = threshold_li(image)
+    >>> binary = image <= thresh
+    """
+    # Requires positive image ( log(mean))
```
remove space after `(`

blink1073 commented on skimage/filters/thresholding.py:
```diff
@@ -302,3 +303,74 @@ def threshold_isodata(image, nbins=256, return_all=False):
+    """
+    # Requires positive image ( log(mean))
+    offset = image.min()
+    # Can't use fixed tolerance for float image
```
Please use `can not`.

blink1073 commented on skimage/filters/thresholding.py:
```diff
@@ -302,3 +303,74 @@ def threshold_isodata(image, nbins=256, return_all=False):
+        old_thresh = new_thresh
+        threshold = old_thresh + tolerance   # range
+        # Calculate the means of background and object pixels
+        # Background
```
I think this and the next comment are just adding confusion.  With the previous comment and the names of the variables, I think you are covered.

blink1073 commented on skimage/filters/thresholding.py:
```diff
@@ -302,3 +303,74 @@ def threshold_isodata(image, nbins=256, return_all=False):
+    # Requires positive image ( log(mean))
+    offset = image.min()
+    # Can't use fixed tolerance for float image
+    imrange = image.max()-offset
```
Please add spaces around `-` operator.

blink1073 commented:
LGTM, thanks @jmetz!

stefanv commented on skimage/filters/thresholding.py:
```diff
@@ -302,3 +303,71 @@ def threshold_isodata(image, nbins=256, return_all=False):
+           http://citeseer.ist.psu.edu/sezgin04survey.html
+    .. [4] ImageJ AutoThresholder code, http://fiji.sc/wiki/index.php/Auto_Threshold
+
+    Adapted for skimage by J. Metz from ImageJ plugin by G.Landini
```
This should be in CONTRIBUTORS.txt

jmetz replied:
@stefanv : I just moved the contribution statement to CONTRIBUTORS.txt and updated the docstring - should I create a new pull request or should this one be reopened?

blink1073 replied:
@jmetz, please open a new PR.

jmetz replied:
Updates now in #1392
